packages/devtools/lib/src/timeline/frame_flame_chart.dart

```diff
@@ -169,37 +200,184 @@ class FrameFlameChart extends CoreElement {
         ..fontWeight = 'bold'
         ..left = '0'
         ..top = '0';
-      section.add(sectionTitle);
+      _gpuSection.add(sectionTitle);
 
-      drawRecursively(frame.gpuEventFlow, 0, section);
+      drawRecursively(_frame.gpuEventFlow, 0, _gpuSection);
     }
 
     drawCpuEvents();
     drawGpuEvents();
+    _setSectionWidths();
+
+    // Calculate our [_minZoomLevel]. At [_zoomLevel] = [_minZoomLevel], the
+    // entire flame chart should fit on the screen.
+    if (element.clientWidth == element.scrollWidth) {
+      _minZoomLevel = 1;
+    } else {
+      _minZoomLevel =
+          min(element.clientWidth / element.scrollWidth * minZoomMultiplier, 1);
+    }
   }
 
   void _drawFlameChartItem(
-      TimelineEvent event, int left, int width, int top, CoreElement section) {
-    final item = Element.div()..className = 'flame-chart-item';
-    final labelWrapper = Element.div()
-      ..className = 'flame-chart-item-label-wrapper';
-    labelWrapper.append(Element.span()
-      ..text = event.name
+    TimelineEvent event,
+    int left,
+    int width,
+    int top,
+    CoreElement section,
+  ) {
+    final item = FlameChartItem(
+      event,
+      left,
+      width,
+      top,
+      event.isCpuEvent ? nextCpuColor() : nextGpuColor(),
+    );
+
+    _chartItems.add(item);
+    section.element.append(item.e);
+  }
+
+  void _setSectionWidths() {
+    final width = getFlameChartWidth();
+    _cpuSection.element.style.width = '${width}px';
+    _gpuSection.element.style.width = '${width}px';
+  }
+
+  num getFlameChartWidth() {
+    int maxRight = 0;
+    for (FlameChartItem item in _chartItems) {
+      if ((item.currentLeft + item.currentWidth) > maxRight) {
+        maxRight = item.currentLeft + item.currentWidth;
+      }
+    }
+    return maxRight;
+  }
+
+  void _handleMouseMove(MouseEvent e) {
+    // Subtract offset so that [currentMouseX] reflects the x coordinate within
+    // the bounds of the flame chart.
+    _currentMouseX = e.client.x - element.offsetLeft;
+
+    // Store current scroll values for re-calculating scroll location on zoom.
+    _currentScrollTop = element.scrollTop;
+    _currentScrollLeft = element.scrollLeft;
+    _currentScrollWidth = element.scrollWidth;
+  }
+
+  void _handleMouseWheel(WheelEvent e) {
+    e.preventDefault();
+
+    if (e.deltaY.abs() > e.deltaX.abs()) {
+      _handleZoom(e.deltaY);
+    } else if (e.deltaX.abs() > e.deltaY.abs()) {
+      // Manually perform horizontal scrolling.
+      element.scrollLeft += e.deltaX;
+      _currentScrollLeft = element.scrollLeft;
+    }
+  }
+
+  void _handleZoom(num deltaY) {
+    if (deltaY > -0.0) {
+      _zoomIn();
```

Comment:
you should use the deltaY to scale how much zoom occurs.

Reply:
added a todo.